java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelReader.java

```diff
@@ -61,13 +61,18 @@ public Message readNextMessage() throws IOException {
       return null;
     }
 
-    buffer = ByteBuffer.allocate(messageLength);
+    return loadMessage(messageLength, ByteBuffer.allocate(messageLength));
+  }
+
+  /**
+   * Read a Message of the given length into the existing buffer and return the loaded Message.
+   */
+  protected Message loadMessage(int messageLength, ByteBuffer buffer) throws IOException {
```

Comment:
Bryan, can you give a bit context why do you want to provide your own buffer when reading the message?

Reply:
So `loadMessage` here will create an Arrow `Message` of the given `messageLength` from the `buffer`.  What I would like to be able to do is override this method to then have access to the message buffer.  This way I can examine the message contents and then copy it to another buffer without having to serialize it again.  For example (in Scala):

```scala
override protected def loadMessage(messageLength: Int, buffer: ByteBuffer): Message = {
        val msg = super.loadMessage(messageLength, buffer)
        val bodyLength = msg.bodyLength().asInstanceOf[Int]

        if (msg.headerType() == MessageHeader.RecordBatch) {
          val allbuf = ByteBuffer.allocate(4 + messageLength + bodyLength)
          allbuf.put(WriteChannel.intToBytes(messageLength))
          allbuf.put(buffer)
          in.readFully(allbuf)
          lastBatch = allbuf.array()
        } else {...}

        msg
```

this allows me to create a new buffer with the existing message and then read the record batch body directly into the new buffer.  So I end up with a single buffer containing the message + body with minimal processing/copying.

Reply:
I see, you want to serialize an arrow record batch message (metadata + body) into a Java on-heap memory rather than Arrow off-heap memory?

Reply:
I think your loadMessage message also changes the semantics of the function from "reading metadata" to "reading metadata and body" in the subclass...

I am not sure it's a good idea from an OOP perspective..

Reply:
>I see, you want to serialize an arrow record batch message (metadata + body) into a Java on-heap memory rather than Arrow off-heap memory?

Sort of.. I need to read a Arrow record batch message from an input stream to a byte array.  I do not need to look at the data and must pass on ownership of the memory,  so I want to avoid loading to Arrow buffers or making unneeded copies.

> I think your loadMessage message also changes the semantics of the function from "reading metadata" to "reading metadata and body" in the subclass...
I am not sure it's a good idea from an OOP perspective..

Nothing changes in `MessageChannelReader`, and the details of a subclass to it aren't really relevant here.  From an Arrow perspective, these classes should provide utility to process streams in a flexible way.  I'm up for any changes if you have a suggestion that is just as efficient and can continue to work with ArrowStreamReader.

Reply:
@BryanCutler how about we discuss on https://github.com/apache/spark/pull/21546/files/a5a1fbe7121c5b0dd93876a56c29ad17dcd9b168#r194898793 first and then decide what to do here?